{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.6.0\n"
     ]
    }
   ],
   "source": [
    "#!/bin/env python\n",
    "# Author: Teja Koganti\n",
    "\n",
    "# 01_setup_mafdb.py \n",
    "# This script does the following - \n",
    "#    1. Reads the MAF file as chunks and for each chunk -\n",
    "            # Calculates VAF\n",
    "            # Filters based on variant_classification columm\n",
    "            # Uses oybedtools to onlyc use variants overlapping target\n",
    "#   2. Establishes a connection to the database and write a table names mutect2 \n",
    "#\n",
    "# Note: requires pandas, sqlite and pybedtools to be installed, and expects python3\n",
    "# conda install -c bioconda pybedtools\n",
    "# conda install -c anaconda pandas\n",
    "# conda install -c anaconda sqlite\n",
    "\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sqlite3\n",
    "import pybedtools\n",
    "import sys\n",
    "\n",
    "\n",
    "# Will be using argparse when the logic looks okay after first PR\n",
    "#maf_file = \"/Users/kogantit/Documents/git_repos/d3b-bix-analysis-toolkit/analyses/TMBanalysis/inputs/temp.maf\"\n",
    "#maf_file = \"/Users/kogantit/Documents/TMB/pbta-snv-mutect2.vep.maf\"\n",
    "maf_file = \"/Users/kogantit/Documents/TMB/PNOC008_MAFfiles/PNOC008_WXS_mergedMAF.maf\"\n",
    "db_name= \"/Users/kogantit/Documents/git_repos/d3b-bix-analysis-toolkit/analyses/TMBanalysis/output/PNOC008WXS_var_db.sqlite\"\n",
    "#db_name = \"/Users/kogantit/Documents/git_repos/d3b-bix-analysis-toolkit/analyses/TMBanalysis/output/var_db.sqlite\"\n",
    "cds_bed=\"/Users/kogantit/Documents/TMB/gencode.v34.annotation.exononly.bed\"\n",
    "#cds_bed = \"/Users/kogantit/Documents/git_repos/d3b-bix-analysis-toolkit/analyses/TMBanalysis/inputs/CCDS_gencode27.bed\"\n",
    "target_bed=\"/Users/kogantit/Documents/TMB/xgen-exome-research-panel-targets_hg38_ucsc_liftover.100bp_padded.sort.merged.bed\"\n",
    "#db_table_name = \"mutect2\"\n",
    "db_table_name = \"consensus_WXS\"\n",
    "#db_intersectbed_table_name = \"mutect2_intersectbed\"\n",
    "db_intersectbed_table_name = \"consensus_WXS_intersectbed\"\n",
    "intersected_bed_out = \"/Users/kogantit/Documents/git_repos/d3b-bix-analysis-toolkit/analyses/TMBanalysis/output/xgen_intersected_cds.bed\"\n",
    "\n",
    "\n",
    "# Function to establish a database connection; else return error message \n",
    "def create_connection(db_file):\n",
    "    \"\"\" create a database connection to a SQLite database \"\"\"\n",
    "    conn = None\n",
    "    try:\n",
    "        conn = sqlite3.connect(db_file)\n",
    "        print(sqlite3.version)\n",
    "    except Error as e:\n",
    "        print(\"Error connecting to database\")\n",
    "    return(conn) \n",
    "\n",
    "\n",
    "def calculate_intersect_BED(target_BED, CDS_bed):\n",
    "    intersect_bed = pybedtools.BedTool(target_BED).intersect(CDS_bed, u=True)\n",
    "    return(intersect_bed)\n",
    "\n",
    "\n",
    "\n",
    "########################################################\n",
    "###########Defining MAF columns and types###############\n",
    "#\n",
    "#\n",
    "# Datatype for each MAF column thta is being used \n",
    "maf_types = [\n",
    "    ('Hugo_Symbol', 'TEXT'),\n",
    "    ('Entrez_Gene_Id', 'INTEGER'),\n",
    "    ('Center', 'TEXT'),\n",
    "    ('NCBI_Build', 'TEXT'),\n",
    "    ('Chromosome', 'TEXT'),\n",
    "    ('Start_Position', 'INTEGER'),\n",
    "    ('End_Position', 'INTEGER'),\n",
    "    ('Variant_Classification', 'TEXT'),\n",
    "    ('Variant_Type', 'TEXT'),\n",
    "    ('Reference_Allele', 'TEXT'),\n",
    "    ('Tumor_Seq_Allele1', 'TEXT'),\n",
    "    ('Tumor_Seq_Allele2', 'TEXT'),\n",
    "    ('dbSNP_RS', 'TEXT'),\n",
    "    ('Tumor_Sample_Barcode', 'TEXT'),\n",
    "    ('Transcript_ID', 'TEXT'),\n",
    "    ('Exon_Number', 'TEXT'),\n",
    "    ('t_depth', 'INTEGER'),\n",
    "    ('t_ref_count', 'INTEGER'),\n",
    "    ('t_alt_count', 'INTEGER'),\n",
    "    ('n_depth', 'INTEGER'),\n",
    "    ('n_ref_count', 'INTEGER'),\n",
    "    ('n_alt_count', 'INTEGER'),\n",
    "    ('Allele', 'TEXT'),\n",
    "    ('flanking_bps', 'TEXT'),\n",
    "    ('VAF', 'REAL')\n",
    "]\n",
    "\n",
    "# Number of lines that read in each chunk from MAF\n",
    "chunksize = 1e5\n",
    "needed_cols = [\n",
    "    'Chromosome',\n",
    "    'Start_Position',\n",
    "    'End_Position',\n",
    "    'Variant_Classification',\n",
    "    'Variant_Type',\n",
    "    'Reference_Allele',\n",
    "    'Tumor_Seq_Allele1',\n",
    "    'Tumor_Seq_Allele2',\n",
    "    'dbSNP_RS',\n",
    "    'Tumor_Sample_Barcode',\n",
    "    'Transcript_ID',\n",
    "    't_depth',\n",
    "    't_ref_count',\n",
    "    't_alt_count',\n",
    "    'n_depth',\n",
    "    'n_ref_count',\n",
    "    'n_alt_count',\n",
    "    'Allele',\n",
    "    'flanking_bps',\n",
    "    'VAF']\n",
    "\n",
    "#Variants will be filtered based on variant_classification \n",
    "#   column if they have these column values \n",
    "# Based on \"Friends of Cancer Research TMB Harmonization Project paper\"\n",
    "#  \"https://jitc.bmj.com/content/8/1/e000147#DC1\"\n",
    "var_class = [\n",
    "    \"Missense_Mutation\",\n",
    "    \"Nonsense_Mutation\",\n",
    "    \"Frame_Shift_Del\",\n",
    "    \"Frame_Shift_Ins\",\n",
    "    \"In_Frame_Del\",\n",
    "    \"In_Frame_Ins\"\n",
    "]\n",
    "\n",
    "# Matching up needed_cols and maf_types \n",
    "needed_types = [col for col in maf_types if col[0] in needed_cols]\n",
    "# Joinging tuples with space in between\n",
    "maf_table_def = \", \".join([\" \".join(col) for col in needed_types])\n",
    "#####################################################\n",
    "\n",
    " \n",
    "\n",
    "    \n",
    "#####################################################\n",
    "########Creating a database with two tables##########\n",
    "#       One where MAF variants intersect with target region\n",
    "#       Second where  MAF variants intersect with target+CDS regions\n",
    "#\n",
    "#\n",
    "# Starting a database connection     \n",
    "con = create_connection(db_name)\n",
    "# Creating a table with name as mutect2, index as an integer and \n",
    "#    maf_table_def as the column names and datatypes\n",
    "\n",
    "#maf_create = \"CREATE TABLE {}('index' INTEGER, {})\".format(\n",
    "#        \"mutect2\", maf_table_def)\n",
    "\n",
    "maf_create = \"CREATE TABLE {}('index' INTEGER, {})\".format(\n",
    "        db_table_name, maf_table_def)\n",
    "maf_intersectbed_create = \"CREATE TABLE {}('index' INTEGER, {})\".format(\n",
    "        db_intersectbed_table_name, maf_table_def)\n",
    "con.execute(maf_create)\n",
    "con.execute(maf_intersectbed_create)\n",
    "####################################################\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#####################################################\n",
    "####Filling up the tables in DB with MAF variants-###\n",
    "#\n",
    "# Creating chunks from MAF file \n",
    "maf_chunks = pd.read_table(maf_file,\n",
    "                               na_values=[\".\"],\n",
    "                               comment=\"#\",\n",
    "                               sep=\"\\t\",\n",
    "                               chunksize=chunksize)\n",
    "\n",
    "intersected_bed = calculate_intersect_BED(target_bed, cds_bed)\n",
    "#merged_intersected_bed = intersected_bed.pybedtools.BedTool.merge()\n",
    "pybedtools.BedTool.to_dataframe(intersected_bed.merge()).to_csv(\n",
    "    intersected_bed_out, sep=\"\\t\", header=None, index=False)\n",
    "\n",
    "for chunk in maf_chunks:\n",
    "    chunk['VAF'] = (chunk['t_alt_count'] /\n",
    "                        (chunk['t_ref_count'] + chunk['t_alt_count']))\n",
    "    chunk = chunk[needed_cols]\n",
    "    chunk_filtered = chunk.loc[chunk.apply(lambda x: \n",
    "                        x[\"Variant_Classification\"]  in var_class, axis=1)]\n",
    "    chunk_within_exon = pybedtools.BedTool.from_dataframe(chunk_filtered).intersect(intersected_bed,\n",
    "                                                                                    u=True)\n",
    "    chunk_within_target = pybedtools.BedTool.from_dataframe(chunk_filtered).intersect(target_bed, u=True)\n",
    "    chunk_target = pybedtools.BedTool.to_dataframe(chunk_within_target, names=needed_cols)\n",
    "    chunk_target.to_sql(db_table_name, con, if_exists='append')\n",
    "    chunk_intersectbed = pybedtools.BedTool.to_dataframe(chunk_within_exon, names=needed_cols)\n",
    "    chunk_intersectbed.to_sql(db_intersectbed_table_name, con, if_exists='append')\n",
    "\n",
    "con.close()\n",
    "#####################################################\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "drop = \"DROP TABLE mutect2\"\n",
    "delete_table=con.execute(drop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
